Core/GCUniqueCommitCollection.m

```diff
@@ -0,0 +1,71 @@
+//  Copyright (C) 2015 Pierre-Olivier Latour <info@pol-online.net>
+//
+//  This program is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU General Public License for more details.
+//
+//  You should have received a copy of the GNU General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+#import "GCUniqueCommitCollection.h"
+
+#import "GCCommit.h"
+
+
+@implementation GCUniqueCommitCollection {
+  NSMutableArray* _commits; // Contains all the commits, even removed ones
+  NSMutableSet* _actualCommitHashes; // Commits that were added but have not been removed
+  NSMutableSet* _removedCommitHashes;
+}
+
+- (instancetype)init {
+  if ((self = [super init])) {
+    _commits = [NSMutableArray array];
+    _actualCommitHashes = [NSMutableSet new];
+    _removedCommitHashes = [NSMutableSet new];
+  }
+  return self;
+}
+
+- (void)addCommit:(GCCommit *)commit {
+  if (![self hasCommit:commit]) {
+    if ([_removedCommitHashes containsObject:commit.SHA1]) {
+      [_removedCommitHashes removeObject:commit.SHA1];
+    } else {
+      [_commits addObject:commit];
+    }
+    [_actualCommitHashes addObject:commit.SHA1];
+  }
+}
+
+- (void)removeCommit:(GCCommit*)commit {
+  if (![self hasCommit:commit]) {
+    return;
+  }
+  // Removing commit object from NSMutableArray is expensive,
+  // so we just moving SHA from one set to another.
+  [_actualCommitHashes removeObject:commit.SHA1];
+  [_removedCommitHashes addObject:commit.SHA1];
+}
+
+- (BOOL)hasCommit:(GCCommit*)commit {
+  return [_actualCommitHashes containsObject:commit.SHA1];
+}
+
+- (NSArray*)commits {
+  NSMutableArray* result = [NSMutableArray array];
+  for (GCCommit* commit in _commits) {
+    if ([self hasCommit:commit]) { // Return only commits that were not removed
```

Comment:
What's the reason for it?
For me `[self hasCommit:commit]` is a shorter and more maintainable thing than using `[_actualCommitHashes containsObject:commit.SHA1]`. 
Like if we want to change the "backend" store for this class and replace `NSMutableSet` if will be less places to make the same changes.
(Maybe I'm overthinking it).